pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -447,19 +457,46 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 	// check the status of containers.
 	for index, container := range pod.Spec.Containers {
 		containerStatus := podStatus.FindContainerStatusByName(container.Name)
-		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
+		containerRef, err := kubecontainer.GenerateContainerRef(pod, &container)
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q, container %v: '%v'", format.Pod(pod), container.Name, err)
+		}
+
+		// Was this container running the last time we checked?
+		wasRunning := false
+		if existingStatus != nil {
+			previousContainerStatus := kubecontainer.GetContainerStatus(existingStatus, container.Name)
+			if previousContainerStatus != nil && previousContainerStatus.State.Running != nil {
+				wasRunning = true
+			}
+		}
+
+		if containerStatus == nil {
+			// No container exists by this name, so just start it
+			changes.ContainersToStart[index] = "Starting container"
+			continue
+		} else if containerStatus.State != kubecontainer.ContainerStateRunning {
+			if wasRunning {
+				// The container was running but currently is not. Record that it exited.
+				exitMessage := fmt.Sprintf("Container exited with status %d", containerStatus.ExitCode)
+				if containerStatus.ExitCode == 0 {
+					m.recorder.Eventf(containerRef, v1.EventTypeNormal, events.ExitSuccess, exitMessage)
```

Comment:
Same response as above, including not wanting to list the container (the event's InvolvedObject is already the container's name)